packages/nitro-protocol/contracts/ninja-nitro/SingleChannelAdjudicator.sol

```diff
@@ -0,0 +1,476 @@
+    * @param turnNumB turnNum of the later state of the pair.
+    * @param appDefinition Address of deployed contract containing application-specific validTransition function.
+    * @return true if the later state is a validTransition from its predecessor, reverts otherwise.
+    */
+    function validTransition(
+        uint256 nParticipants,
+        bool[2] memory isFinalAB, // [a.isFinal, b.isFinal]
+        IForceMoveApp.VariablePart[2] memory ab, // [a,b]
+        uint48 turnNumB,
+        address appDefinition
+    ) public pure returns (bool) {
+        return _requireValidTransition(nParticipants, isFinalAB, ab, turnNumB, appDefinition);
+    }
+
+    // ASSET HOLDING PART
+
+    /**
+     * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
+     * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
+     * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
+     * @param outcomeBytes The abi.encode of AssetOutcome.Allocation
+     * @param indices Array with each entry denoting the index of a destination to transfer funds to. An empty array indicates "all".
+     */
+    function transfer(
+        bytes32 fromChannelId,
+        uint48 turnNumRecord,
+        uint48 finalizesAt,
+        bytes32 stateHash,
+        address challengerAddress,
+        bytes calldata outcomeBytes,
+        uint256[] memory indices
+    ) external {
+        // checks
+        _requireIncreasingIndices(indices);
+
+        bytes32 outcomeHash = keccak256(outcomeBytes);
+        _requireMatchingStorage(
+            ChannelData(turnNumRecord, finalizesAt, stateHash, challengerAddress, outcomeHash),
+            fromChannelId
+        );
+
+        // effects and interactions
+
+        // loop over tokens
+        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
+
+        for (uint256 i = 0; i < outcome.length; i++) {
+            Outcome.AssetOutcome memory assetOutcome = abi.decode(
+                outcome[i].assetOutcomeBytes,
+                (Outcome.AssetOutcome)
+            );
+
+            if (assetOutcome.assetOutcomeType == uint8(Outcome.AssetOutcomeType.Allocation)) {
+                assetOutcome.allocationOrGuaranteeBytes = abi.encode(
+                    _transfer(
+                        outcome[i].assetHolderAddress,
+                        assetOutcome.allocationOrGuaranteeBytes,
+                        indices
+                    )
+                );
+                outcome[i] = Outcome.OutcomeItem(
+                    outcome[i].assetHolderAddress,
+                    abi.encode(assetOutcome)
+                );
+            } else {
+                revert('AssetOutcome not an allocation');
+            }
+        }
+
+        statusOf[fromChannelId] = _generateStatus(
```

Comment:
This unconditionally writes to storage, even when the resulting outcome is empty. At what point does the adjudicator self destruct?

Reply:
The self destruction optimisation is  currently used only in the happy path, whereas the public function `transfer` is part of the unhappy path. The next round of improvements to ninja nitro should extend it to this function, too.

In vanilla nitro we track a `safeToDelete` boolean to cover this kind of thing. Here we could have a `safeToSelfDestruct` flag returned from one of the internal functions. 

Overall it would be nice if the adjudicator _always_ knew when it could self destruct, and did so eagerly if safe.

Reply:
Is it easy to decide when it should self destruct? If either
- the outcome is empty
- the holdings is 0 for every asset

does the adjudicator serve a further purpose?

Reply:
It is easy to decide: it should self destruct only if the outcome is empty: which is to say if there is nothing stored in the adjudicator. It doesn't matter if there are some holdings, since these are not stored in the adjudicator. The adjudicator can always be redeployed (with blank storage).